### PR TITLE
Add check of Dir.home's permission

### PR DESCRIPTION
**Pull request by allenzhao**
Ref #4778

### CODE REVIEW

File: lib/bundler/shared_helpers.rb [hunk 1]
```diff
@@ -136,6 +136,10 @@ def print_major_deprecations!
       major_deprecation("Bundler will only support rubygems >= 2.0, you are running #{Bundler.rubygems.version}")
     end
 
+    def check_home_dir_permissions
+      raise PathError, "There was an error while trying to use your home path #{Dir.home}" unless File.writable?(Dir.home)
```
Comment (segiddins) [inline lib/bundler/shared_helpers.rb]: this should specifically mention that the home directory must exist & be writable

File: spec/bundler/shared_helpers_spec.rb [hunk 2]
```diff
@@ -412,4 +412,11 @@ module TargetNamespace
       end
     end
   end
+
+  describe "#check_home_dir_permissions" do
+    it "raises a PathError" do
+      allow(File).to receive(:writable?).with(Dir.home).and_return(false)
```
Comment (segiddins) [inline spec/bundler/shared_helpers_spec.rb]: please also add a test for when `Dir.home` does not exist

**Comment by segiddins** [comment]
@allenzhao ping?

**Comment by indirect** [comment]
I think Bundler should just work, even if home directories are unwritable or do not exist. User-level settings should not be persisted, and cached files like the compact index and gem download cache should go into `/tmp/bundler` instead.

File: spec/bundler/shared_helpers_spec.rb (hunk 2 above)
Comment (JuanitoFatas) [inline spec/bundler/shared_helpers_spec.rb]: Regarding to `Dir.home`, I found using user's login is more reliable:

```
# suppose ENV HOME is gone
> ENV.delete 'HOME'

> Dir.home
ArgumentError: couldn't find HOME environment -- expanding `~'

> require 'etc'; File.expand_path(Dir.home(Etc.getlogin))
> "/Users/Juan"
```

**Comment by allenzhao** [comment]
@indirect So what should be the solution here?
@segiddins I've already added the test and updated according to @JuanitoFatas 's advice.
Thanks!

**Comment by segiddins** [comment]
@allenzhao there are RuboCop failures in addition

**Comment by indirect** [comment]
@allenzhao I added notes to the list ticket #4871:

> Use /tmp/bundler for compact_index when Dir.home isn't writable #4792 #4819
> Skip persisting global settings to disk if Dir.home is missing or unwritable #4918

I think it means we will want to fix this problem in a different way, with PRs that solve those two problems. I would be very happy to have you work on those, and if you would like me to write a more detailed description of how that should work, I am happy to do it, just let me know.

**Comment by allenzhao** [comment]
@indirect Please do by opening a new issue, and I will close this PR.

**Pull request closed without merging**